### PR TITLE
Use amqp HandleMessage Func

### LINKED ISSUE #189: Long messages and "Servicebus Lock Duration" (go-amqp)
(this is also registered in https://github.com/Azure/go-amqp/issues/12)

We are experiencing weird behaviour when messages take longer than half the lock-duration of the servicebus queue or topic.

Example:
Queue with a lock-duration of 30 seconds
All messages taking 20 seconds before Accept.

Expected behaviour:
All messages to be read and accepted by the servicebus

Actual behaviour:
1st message gets accepted correctly by the queue
all subsequent messaged are reported to the client as accepted, but the queue handles them like "Release()"
i.e no errors are reported from the "msg.Accept()"-call

The timing is always half of the "lock-duration", i.e setting lock-duration to max 5 minutes, messages taking >2.5 minutes behave the same as the scenario above.
and all messages taking less than half the lock-duration works as they should.

after some extended testing it seems that some message occasionally gets accepted correctly, but most messages end up as "released"

### PR DESCRIPTION
**Pull request by serbrech**
Make use of the new HandleMessage api in go-amqp

This bumps the version of go-amqp to include the following bug fixes: 
- don't over-fetch messages (leading to inevitable lock-expiry of messages)  (https://github.com/Azure/go-amqp/pull/20)
- bubble message disposition error to client (https://github.com/Azure/go-amqp/pull/19)

fixes #189

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,3 +18,5 @@ require (
 	golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd // indirect
 	nhooyr.io/websocket v1.8.6
 )
+
+replace github.com/Azure/go-amqp => ../go-amqp
```
Comment (serbrech) [inline go.mod:22]: obviously, that's not final 🐗

File: receiver.go [hunk 2]
```diff
@@ -216,24 +214,14 @@ func (r *Receiver) Listen(ctx context.Context, handler Handler) *ListenerHandle
 	ctx, span := r.startConsumerSpanFromContext(ctx, "sb.Receiver.Listen")
 	defer span.End()
 
-	messages := make(chan *amqp.Message)
-	go r.listenForMessages(ctx, messages)
-	go r.handleMessages(ctx, messages, handler)
```
Comment (serbrech) [inline receiver.go:221]: the handleMessages is now folded into the `amqpAdapterHandler` that wraps the user handler

File: receiver.go [hunk 3]
```diff
@@ -216,24 +214,14 @@ func (r *Receiver) Listen(ctx context.Context, handler Handler) *ListenerHandle
 	ctx, span := r.startConsumerSpanFromContext(ctx, "sb.Receiver.Listen")
 	defer span.End()
 
-	messages := make(chan *amqp.Message)
```
Comment (serbrech) [inline receiver.go:219]: no more channel to manage. we don't need an implicit buffer. prefetch messages are already buffered in `go-amqp`

File: receiver.go [hunk 4]
```diff
@@ -216,24 +214,14 @@ func (r *Receiver) Listen(ctx context.Context, handler Handler) *ListenerHandle
...
-	go r.handleMessages(ctx, messages, handler)
+	go r.listenForMessages(ctx, newAmqpAdapterHandler(r, handler))
 
 	return &ListenerHandle{
 		r:   r,
 		ctx: ctx,
 	}
 }
 
-func (r *Receiver) handleMessages(ctx context.Context, messages chan *amqp.Message, handler Handler) {
```
Comment (serbrech) [inline receiver.go:229]: no need since we pass it down to the handler without the buffer channel

File: receiver.go [hunk 5]
```diff
@@ -345,29 +330,25 @@ func (r *Receiver) setLastError(err error) {
...
 	var receiver *amqp.Receiver
 	r.clientMu.RLock()
 	if r.receiver == nil {
-		return nil, r.connClosedError(ctx)
+		return r.connClosedError(ctx)
 	}
 	receiver = r.receiver
 	r.clientMu.RUnlock()
-	msg, err := receiver.Receive(ctx)
+	err := receiver.HandleMessage(ctx, func(message *amqp.Message) error {
```
Comment (serbrech) [inline receiver.go:344]: call HandleMessage instead of Receive which we deprecate

File: go.mod [hunk 6]
```diff
@@ -3,7 +3,7 @@ module github.com/Azure/azure-service-bus-go
 go 1.12
 
 require (
-	github.com/Azure/azure-amqp-common-go/v3 v3.0.1
+	github.com/Azure/azure-amqp-common-go/v3 v3.1.0
```
Comment (serbrech) [inline go.mod:6]: actually not sure what changed here?

File: receiver.go [hunk 7]
```diff
@@ -345,29 +330,25 @@ func (r *Receiver) setLastError(err error) {
...
+		return handler.Handle(ctx, message)
+	})
 	if err != nil {
 		tab.For(ctx).Debug(err.Error())
-		return nil, err
-	}
-
-	id := messageID(msg)
-	if idStr, ok := id.(string); ok {
-		span.AddAttributes(tab.StringAttribute("amqp.message.id", idStr))
```
Comment (jhendrixMSFT) [inline receiver.go:367]: We don't need this now?

Comment (jhendrixMSFT) [reply receiver.go:367]: Oops missed it.

File: amqphandler.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
+package servicebus
```
Comment (jhendrixMSFT) [inline amqphandler.go:1]: Missing MIT license header.

File: receiver.go (hunk 7 above)
Comment (serbrech) [reply receiver.go:367]: right, the whole internal listenForMessage is now folded into the middlware handler that does the translation from amqp to SB message type.

File: amqphandler.go (hunk 8 above)
Comment (serbrech) [reply amqphandler.go:1]: fixed

File: go.mod [hunk 9]
```diff
@@ -3,14 +3,13 @@ module github.com/Azure/azure-service-bus-go
...
 require (
-	github.com/Azure/azure-amqp-common-go/v3 v3.0.1
-	github.com/Azure/azure-sdk-for-go v37.1.0+incompatible
-	github.com/Azure/go-amqp v0.13.1
-	github.com/Azure/go-autorest/autorest v0.11.7
-	github.com/Azure/go-autorest/autorest/adal v0.9.4
+	github.com/Azure/azure-amqp-common-go/v3 v3.1.0
+	github.com/Azure/azure-sdk-for-go v51.1.0+incompatible
+	github.com/Azure/go-amqp v0.13.3
+	github.com/Azure/go-autorest/autorest v0.11.18
```
Comment (serbrech) [inline go.mod:9]: should I keep these version bumps for another PR?

Comment (jhendrixMSFT) [reply go.mod:9]: The go-autorest and azure-sdk-for-go should be fine as they're used for tests.

Updating common might be useful as it adds support for managed service identity which people have been asking for.  I don't think simply makes that work but it's a step in the right direction.

Comment (serbrech) [reply go.mod:9]: I have the msi implementation in go-shuttle, so we can contribute it back here

**Comment by serbrech** [comment]
I would prefer this PR to merge in go-amqp : https://github.com/Azure/go-amqp/pull/23
I can then update this PR to point at the latest go-amqp (new patch release after above PR merges)

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by Segflow** [comment]
Using this version instead of `0.10.8` broke our services. Seems like using amqp `HandleMessage` do something different that what was used before this. 

More info here: https://github.com/google/go-cloud/issues/2980

**Comment by jhendrixMSFT** [comment]
@Segflow are you calling a disposition action after processing the message?  This is a requirement per the AMQP protocol.  Earlier versions didn't honor this and caused the problems described in https://github.com/Azure/go-amqp/pull/20

**Comment by Segflow** [comment]
`gocloud.dev` fetch the messages and store them in a slice, and return from the handler. 

Later the user decides if they want to Ack or Nack the message.

**Comment by serbrech** [comment]
this was the implementation before, and it did not allow to fulfil flow control in AMQP, which needs to communicate back to the upstream peer.
this is now respecting flow control.
The problem is that your application does not set prefetch, so it's at 1.
as long as no message disposition is taken, the server won't send you the next one.
